Add XDELEX and XACKDEL commands for stream

sundb opened the pull request.

tacheraSasi reviewed (approved):
looks good to me

tacheraSasi reviewed (approved):
Lgtm

moticless commented:
Hi @sundb, 
I am not much familiar with stream low level details and doing my baby steps so I might say inaccurate things. 
* Can we avoid from yet another data structure to maintain, if we could have attach per entry in the stream some metadata. And in this metadata will embed the logic of cgroups_ref.
* How many consumer groups usually users have? Is ACKED expected to be more common than update of the linked list? Maybe it is better to start simple, without additional data structure and just iterate over all the consumer groups.

sundb commented:
> * Can we avoid from yet another data structure to maintain, if we could have attach per entry in the stream some metadata. And in this metadata will embed the logic of cgroups_ref.

We can't do this because we might delete entry but not delete the references, so they can't be together.

> * How many consumer groups usually users have? Is ACKED expected to be more common than update of the linked list? Maybe it is better to start simple, without additional data structure and just iterate over all the consumer groups.

I changed to store a minimum last_id, invalidate it when modifying the group, and update it when in use.
now the code is simpler.
[`5055b1f` (#14130)](https://github.com/redis/redis/pull/14130/commits/5055b1fb20e55b2fee9f706e5ed541319eee89da)

tezc commented on src/t_stream.c:
```diff
@@ -3608,6 +3988,101 @@ void xdelCommand(client *c) {
+ * - DELREF: Remove messages and clean up all consumer group references
+ * - ACKED: Only remove messages that have no pending consumers */
+void xdelexCommand(client *c) {
+    kvobj *kv = lookupKeyWriteOrReply(c, c->argv[1], shared.czero); 
```
if key does not exist, it will return zero right? I feel like it should've been XDELEX_NO_ID, no? Also, I don't see if we ever return XDELEX_NO_ID? Is this intentional?

sundb replied:
XDELEX_NO_ID does not apply to it. Just like xdel, if the key dones't exist, we return 0.

tezc replied:
okay, initially I missed, this command is supposed to return "array of results". Then, we return array of results if the key exists, otherwise, we don't return array but just zero. I feel like this is unlike what we have been doing with other commands. There were similar cases and we changed all to return array no matter what e.g. hmget, httl etc. 
If this is intentional, that's ok. 

Still, we don't have a case that it will return `XDELEX_NO_ID`. We document to return it but don't see where do we return it.

sundb replied:
i'm sure what is correct, let's wait for @moticless 's opinion.

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/e1f0adab615a07ec8bbd155b45d3a5e198c0ad87

tezc commented on src/t_stream.c:
```diff
@@ -3608,6 +3988,101 @@ void xdelCommand(client *c) {
+/* XDELEX <key> [KEEPREF|DELREF|ACKED] [IDS <numids> <id ...>]
+ *
+ * Removes the specified entries from the stream. Returns the number
+ * of items actually deleted, that may be different from the number
```
I think it does not return number of deleted items right?

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/f3a5948e6932a81fe9b32dd93dc4c41f19151f44

skaslev commented on src/stream.h:
```diff
@@ -21,6 +21,9 @@ typedef struct stream {
     uint64_t entries_added; /* All time count of elements added. */
     rax *cgroups;           /* Consumer groups dictionary: name -> streamCG */
+    rax *cgroups_ref;       /* Index mapping message IDs to their consumer groups. */
+    streamID min_cgroup_last_id;  /* The minimum ID of consume groupw. */
```
typo: `consume groupw` -> `consumer group`

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/3f9aac3a91c77f6c74312e3d4ab41b2460188c37

skaslev commented on src/t_stream.c:
```diff
@@ -964,6 +1021,12 @@ static int streamParseAddOrTrimArgsOrReply(client *c, streamAddTrimArgs *args, i
             }
             limit_given = 1;
             i++;
+        } else if (!strcasecmp(opt,"keepref") && args->delete_strategy == DELETE_STRATEGY_NONE) {
+            args->delete_strategy = DELETE_STRATEGY_KEEPREF;
+        } else if (!strcasecmp(opt,"delref") && args->delete_strategy == DELETE_STRATEGY_NONE) {
+            args->delete_strategy = DELETE_STRATEGY_DELREF;
+        } else if (!strcasecmp(opt,"acked") && args->delete_strategy == DELETE_STRATEGY_NONE) {
+            args->delete_strategy = DELETE_STRATEGY_ACKED;
```
maybe factor the common check out?

```
       } else if (args->delete_strategy == DELETE_STRATEGY_NONE) {
            if (!strcasecmp(opt,"keepref")) {
                args->delete_strategy = DELETE_STRATEGY_KEEPREF;
            } else if (!strcasecmp(opt,"delref")) {
                args->delete_strategy = DELETE_STRATEGY_DELREF;
            } else if (!strcasecmp(opt,"acked")) {
                args->delete_strategy = DELETE_STRATEGY_ACKED;
            }
       } else if (...
```

sundb replied:
if so, we can't enter into `else if (...` if the argument is not keepref, delref or acked.

tezc commented on tests/unit/type/stream-cgroups.tcl:
```diff
@@ -543,6 +543,25 @@ start_server {
         assert_equal [s total_error_replies] 1
     }
 
+    test {XGROUP DESTROY removes all consumer group references} {
```
if I'm not mistaken, XTRIM test with new options is missing. Does it make sense to add it just for coverage or maybe we hit same underlying functions enough from other commands?

sundb replied:
aha, yes, `XTRIM` and `XADD` share this trim code.

tezc commented on src/t_stream.c:
```diff
@@ -2901,6 +3166,120 @@ void xackCommand(client *c) {
+ * - It acknowledges messages in the Pending Entries List (PEL) like XACK
+ * - It also deletes the acknowledged messages from the stream like XDEL
+ * 
+ * Return value is the number of messages successfully acknowledged. */
```
I guess return value comment is outdated here as well

sundb replied:
Done with https://github.com/redis/redis/pull/14130/commits/0e6abccfe90a1804793847caf544c8a2afffd6aa
and i removed some comment that duplicated with ELETE_STRATEGY_*

tezc commented on src/t_stream.c:
```diff
@@ -2901,6 +3166,120 @@ void xackCommand(client *c) {
+
+    /* No key or group? Nothing to ack. */
+    if (kv == NULL || group == NULL) {
+        addReply(c,shared.czero);
```
I feel same way here, feel like it could be an array.

sundb replied:
@moticless There was once a similar fix that changed the return integer value to an empty array, but I forgot which one it was.

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/e1f0adab615a07ec8bbd155b45d3a5e198c0ad87

tezc replied:
forgot to delete this if check?

sundb replied:
bind me.

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/ea94241fc9bfd359670f9d303a44771e83c26fd7

skaslev commented on src/t_stream.c:
```diff
@@ -3608,6 +3985,95 @@ void xdelCommand(client *c) {
+    int deleted = 0;
+    addReplyArrayLen(c, args.numids);
+    for (int j = 0; j < args.numids; j++) {
+        int res = XDELEX_NO_ID;
```
This default value always gets overwritten below at `res = can_delete ? XDELEX_DELETED : XDELEX_STILL_REFERENCED;`

sundb replied:
good catch.

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/31ef4c1f3a7c9521778036363a82d9dc3e40b2a1 and add a test to cover it.

skaslev commented on src/t_stream.c:
```diff
@@ -3608,6 +3985,95 @@ void xdelCommand(client *c) {
+        }
+    }
+
+    /* Propagate the write if needed. */
```
nit: fold in the above `if (deleted) {`?

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/262b71ff79055dccb273c4c87e5492c8681584de
and remove a duplicate server.dirty increment.

tezc commented on src/commands/xackdel.json:
```diff
@@ -0,0 +1,109 @@
+            }
+        ],
+        "reply_schema": {
+            "description": "Array of results. Returns empty array if the key does not exist.",
```
just adding a comment as a reminder to update this line if needed (it does not return empty array)

sundb replied:
aha, i recall it, it needs to return an empty array, where seen originally is https://github.com/redis/redis/pull/13295

sundb replied:
FYI @LiorKogan

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/e1f0adab615a07ec8bbd155b45d3a5e198c0ad87

tezc replied:
reminder to fix this comment. It won't return empty array.

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/d694076c33df1b11fdd4ac71d868bebf7fb19724

tezc commented on src/commands/xdelex.json:
```diff
@@ -0,0 +1,104 @@
+            }
+        ],
+        "reply_schema": {
+            "description": "Array of results. Returns empty array if the key does not exist.",
```
just adding a comment as a reminder to update this line if needed (it does not return empty array)

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/e1f0adab615a07ec8bbd155b45d3a5e198c0ad87

tezc replied:
reminder to fix this comment. It won't return empty array.

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/d694076c33df1b11fdd4ac71d868bebf7fb19724

tezc commented on src/commands/xackdel.json:
```diff
@@ -0,0 +1,109 @@
+                "oneOf": [
+                    {
+                        "description": "The id does not exist in the provided stream key.",
+                        "const": -2
```
btw any reason why we don't use -1?

sundb replied:
I just wanted to synchronize with the command of hfe. It does seem strange now.

sundb replied:
done with [`db8c437` (#14130)](https://github.com/redis/redis/pull/14130/commits/db8c437de49d0e776c5b88a0cbbc9d328a7b6d2b)

tezc commented on src/t_stream.c:
```diff
@@ -3608,6 +3988,106 @@ void xdelCommand(client *c) {
+    if (!kv) {
+        addReplyArrayLen(c, args.numids);
+        for (int i = 0; i < args.numids; i++)
+            addReplyNull(c);
```
can't we reply with XDELEX_NO_ID? or maybe some other integer XDELEX_NO_KEY? Because in other similar PRs, we didn't change the type in the array. See `httl` command, it does not reply with nulls.

Edit: Feel like we should return XDELEX_NO_ID? If we can't find the id in the stream, we return XDELEX_NO_ID. So, we can't find it in an empty/non-existing key? I'm not entirely confident in my understanding of how streams work, so please take this with a grain of salt.

sundb replied:
i think we should return XDELEX_NO_ID, when the Key does not exist, we assume that the key is empty instead of non-existent.

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/cf42f935778ace4ed8408657c5b3663c6808330a

ShooterIT commented on src/t_stream.c:
```diff
@@ -1019,9 +1082,65 @@ static int streamParseAddOrTrimArgsOrReply(client *c, streamAddTrimArgs *args, i
+            if (getRangeLongFromObjectOrReply(c, c->argv[j+1], 1, LONG_MAX,
+                &args->numids, "Number of IDs must be a positive integer") != C_OK) {
+                return 0;
+            }
```
```suggestion
            if (getRangeLongFromObjectOrReply(c, c->argv[j+1], 1, LONG_MAX,
                &args->numids, "Number of IDs must be a positive integer") != C_OK)
            {
                return 0;
            }
```

ShooterIT commented on src/t_stream.c:
```diff
@@ -2467,6 +2588,127 @@ void xreadCommand(client *c) {
+    /* When a consumer group's last_id is updated, we need to invalidate the cached
+    * minimum last_id in two cases:
+    * 1. If the consumer group's previous last_id equals the minimum last_id.
+    * 2. If the new ID being set is smaller than the current minimum last_id. */
```
```suggestion
    /* When a consumer group's last_id is updated, we need to invalidate the cached
     * minimum last_id in two cases:
     * 1. If the consumer group's previous last_id equals the minimum last_id.
     * 2. If the new ID being set is smaller than the current minimum last_id. */
```

ShooterIT commented on src/t_stream.c:
```diff
@@ -2901,6 +3166,120 @@ void xackCommand(client *c) {
+                /* Only delete if acknowledged by all consumer groups */
+                if (streamEntryIsReferenced(s, id))
+                    can_delete = 0;
+            } else if (args.delete_strategy == DELETE_STRATEGY_DELREF) {
+                streamCleanupEntryCGroupRefs(s, id);
```
```suggestion
                /* We want to know if the first entry in the stream was deleted
                 * so we can later set the new one. */
```

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/7a18c3aff08c58209d4139f198bca448d798da84

ShooterIT commented on src/t_stream.c:
```diff
@@ -3608,6 +3988,106 @@ void xdelCommand(client *c) {
+     * key does not exist.*/
+    if (!kv) {
+        addReplyArrayLen(c, args.numids);
+        for (int i = 0; i < args.numids; i++)
```
```suggestion
        if (streamParseStrictIDOrReply(c,c->argv[j+args.startidx],&ids[j],0,NULL) != C_OK)
            goto cleanup;
```

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/7a18c3aff08c58209d4139f198bca448d798da84

ShooterIT commented on src/t_stream.c:
```diff
@@ -3608,6 +3988,106 @@ void xdelCommand(client *c) {
+    for (int j = 0; j < args.numids; j++) {
+        int res = XDELEX_NO_ID;
+        streamID *id = &ids[j];
+        unsigned char buf[sizeof(streamID)];
+        streamEncodeID(buf,id);
```
```suggestion
                /* We want to know if the first entry in the stream was deleted
                 * so we can later set the new one. */
```

sundb replied:
done with https://github.com/redis/redis/pull/14130/commits/7a18c3aff08c58209d4139f198bca448d798da84

ShooterIT commented:
from POV of `PEL`, returning -1 means `ID not found in the stream`, other values ​​mean PEL ACK is successful

from POV of entry:
if we use `0` and `1`, i prefer `1` means entry is also deleted. `0` means entry is not deleted. 
if we use `1` and `2`, i prefer `2` means entry is also deleted(both entry and PEL are deleted), `1` means entry is not deleted(only PEL is deleted)

sundb commented:
> from POV of `PEL`, returning -1 means `ID not found in the stream`, other values ​​mean PEL ACK is successful
> 
> from POV of entry: if we use `0` and `1`, i prefer `1` means entry is also deleted. `0` means entry is not deleted. if we use `1` and `2`, i prefer `2` means entry is also deleted(both entry and PEL are deleted), `1` means entry is not deleted(only PEL is deleted)

I’m not opposed to either approach, I just want a consistent rule that future commands can follow. Also, I prefer not to give return codes any meaning beyond indicating errors.
@tezc @moticless @LiorKogan please share your thoughts.

tezc commented:
I guess it is okay not to use zero as success return value. I see we did something similar for HPERSIST.

sundb commented:
fully CI: https://github.com/sundb/redis/actions/runs/15995761819
